File: files-app/src/main/java/l/files/fs/local/LocalResourceStatus.java
package l.files.fs.local;

import java.io.IOException;
import java.util.Set;

import auto.parcel.AutoParcel;
import l.files.fs.Instant;
import l.files.fs.LinkOption;
import l.files.fs.Permission;
import l.files.fs.ResourceStatus;

import static java.util.Objects.requireNonNull;
import static l.files.fs.LinkOption.FOLLOW;
import static l.files.fs.local.LocalResource.mapPermissions;

@AutoParcel
public abstract class LocalResourceStatus implements ResourceStatus {

    private Instant atime;
    private Instant mtime;
    private Set<Permission> permissions;

    public abstract Stat getStat();

    public static LocalResourceStatus create(Stat stat) {
        return new AutoParcel_LocalResourceStatus(stat);
    }

    @Override
    public Instant getAccessTime() {
        if (atime == null) {
            atime = Instant.of(getStat().getAtime(), getStat().getAtimeNsec());
        }
        return atime;
    }

    @Override
    public Instant getModificationTime() {
        if (mtime == null) {
            mtime = Instant.of(getStat().getMtime(), getStat().getMtimeNsec());
        }
        return mtime;
    }

    @Override
    public long getSize() {
        return getStat().getSize();
    }

    @Override
    public boolean isSymbolicLink() {
        return Stat.S_ISLNK(getStat().getMode());
    }

    @Override
    public boolean isRegularFile() {
        return Stat.S_ISREG(getStat().getMode());
    }

    @Override
    public boolean isDirectory() {
        return Stat.S_ISDIR(getStat().getMode());
    }

    public long getDevice() {
        return getStat().getDev();
    }

    public long getInode() {
        return getStat().getIno();
    }

    @Override
    public boolean isFifo() {
        return Stat.S_ISFIFO(getStat().getMode());
    }

    @Override
    public boolean isSocket() {
        return Stat.S_ISSOCK(getStat().getMode());
    }

    @Override
    public boolean isBlockDevice() {
        return Stat.S_ISBLK(getStat().getMode());
    }

    @Override
    public boolean isCharacterDevice() {
        return Stat.S_ISCHR(getStat().getMode());
    }

    @Override
    public Set<Permission> getPermissions() {
        if (permissions == null) {
            permissions = mapPermissions(getStat().getMode());
        }
        return permissions;
    }

    public static LocalResourceStatus stat(
            LocalResource resource,
            LinkOption option) throws IOException {

        requireNonNull(option, "option");
        Stat stat;
        try {
            if (option == FOLLOW) {
                stat = Stat.stat(resource.getPath());
            } else {
                stat = Stat.lstat(resource.getPath());
            }
        } catch (ErrnoException e) {
            throw e.toIOException(resource.getPath());
        }

        return create(stat);
    }

}

File: files-app/src/main/kotlin/fs/local/LocalResourceStatus.kt
package l.files.fs.local

import java.io.File
import java.io.IOException

import l.files.fs.ResourceStatus
import l.files.fs.Path

import kotlin.properties.Delegates
import kotlin.platform.platformStatic
import com.google.common.net.MediaType

private data class LocalResourceStatus private (
        private val _path: LocalPath,
        val stat: Stat) : ResourceStatus {

    override fun getPath() = _path

    override fun getResource() = getPath().getResource()

    override val isReadable: Boolean by Delegates.lazy { access(Unistd.R_OK) }
    override val isWritable: Boolean by Delegates.lazy { access(Unistd.W_OK) }
    override val isExecutable: Boolean by Delegates.lazy { access(Unistd.X_OK) }

    private fun access(mode: Int) = try {
        Unistd.access(getPath().toString(), mode)
        true
    } catch (e: ErrnoException) {
        false
    }

    override val lastModifiedTime = stat.mtime * 1000
    override val size = stat.size

    override val isSymbolicLink = Stat.S_ISLNK(stat.mode)
    override val isRegularFile = Stat.S_ISREG(stat.mode)
    override val isDirectory = Stat.S_ISDIR(stat.mode)

    val device = stat.dev
    val inode = stat.ino

    val isFifo = Stat.S_ISFIFO(stat.mode)
    val isSocket = Stat.S_ISSOCK(stat.mode)
    val isBlockDevice = Stat.S_ISBLK(stat.mode)
    val isCharacterDevice = Stat.S_ISCHR(stat.mode)

    override val basicMediaType: MediaType by Delegates.lazy {
        try {
            BasicFileTypeDetector.detect(this)
        } catch (e: IOException) {
            MediaType.OCTET_STREAM
        }
    }

    public companion object {

        throws(javaClass<IOException>())
        platformStatic fun stat(file: File, followLink: Boolean) =
                stat(LocalPath.of(file), followLink)

        throws(javaClass<IOException>())
        platformStatic fun stat(path: Path, followLink: Boolean): LocalResourceStatus {
            LocalPath.check(path)
            val stat: Stat
            try {
                if (followLink) {
                    stat = Stat.stat(path.toString())
                } else {
                    stat = Stat.lstat(path.toString())
                }
            } catch (e: ErrnoException) {
                throw e.toIOException()
            }

            return LocalResourceStatus(path as LocalPath, stat)
        }
    }
}
